localstack/services/logs/provider.py

```diff
@@ -247,46 +246,61 @@ def moto_put_subscription_filter(fn, self, *args, **kwargs):
     role_arn = args[4]
 
     log_group = self.groups.get(log_group_name)
+    log_group_arn = arns.log_group_arn(log_group_name, self.account_id, self.region_name)
 
     if not log_group:
         raise ResourceNotFoundException("The specified log group does not exist.")
 
+    arn_data = arns.parse_arn(destination_arn)
+
+    if role_arn:
+        factory = connect_to.with_assumed_role(
+            role_arn=role_arn, service_principal=ServicePrincipal.logs
+        )
+    else:
+        factory = connect_to(aws_access_key_id=arn_data["account"], region_name=arn_data["region"])
```

Comment:
q: is this a fallback? this whole `with_assumed_role` addition is looking really nice 😄 I lack some knowledge in the IAM part, so this kind of things I don't feel qualified enough to review.

Reply:
You're correct, if a role ARN is available, IAM enforcement can be performed. Else there is a fallback.

This usage is similar to elsewhere, eg:

https://github.com/localstack/localstack/blob/05832fc4e4b0e23db05a023564986acb2d1d8c8a/localstack/services/sns/publisher.py#L646-L652

or

https://github.com/localstack/localstack/blob/05832fc4e4b0e23db05a023564986acb2d1d8c8a/localstack/services/firehose/provider.py#L734-L739

Reply:
IAM enforcement can be performed in both cases. However, _only_ if the aws_access_key_id is set to the `INTERNAL_AWS_ACCESS_KEY_ID` value. There should be no case, other than resource creation, to set the account id as access key id for internal clients, in my opinion. The first snippet by viren for example leads to no IAM enforcement in the else case, which is neither parity nor should happen.